### PR TITLE
Datasets finder for Data replication 

### PR DESCRIPTION
**Pull request by tuGithub**

### CODE REVIEW

**Comment by tuGithub** [comment]
ConfigBasedDataset is a place holder now

**Comment by tuGithub** [comment]
@chavdar , @ibuenros , please review when you have time

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+      this.replicationDisableTags = Optional.of(disablePaths);
+    }
+    else{
+      this.replicationDisableTags = Optional.absent();
+    }
+
+    configClient = ConfigClient.createConfigClient(VersionStabilityPolicy.WEAK_LOCAL_STABILITY);
+  }
+
+  protected static Set<URI> getValidDatasetURIs(Collection<URI> allDatasetURIs, Collection<URI> disabledURIs, Path ancestor) {
```
Comment (chavdar) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: nit: Can you choose a better name than "ancestor"?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  @Override
+  public List<ConfigBasedDataset> findDatasets() throws IOException {
+    Collection<URI> allDatasetURIs;
+    Collection<URI> disabledURIs = ImmutableList.of();;
+    try {
+      // get all the URIs which imports {@link #replicationTag}
+      allDatasetURIs = configClient.getImportedBy(new URI(replicationTag.toString()), true);
+      
+      if(this.replicationDisableTags.isPresent()){
+        disabledURIs = new ArrayList<URI>();
```
Comment (chavdar) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: You seem to first initialize this as a list and then you convert to a set. You might as well, use a set to begin with.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    Set<URI> validURISet = new HashSet<URI>();
+
+    // remove non leaf URI
+    for (URI u : datasetsList) {
+      URI needToRemove = null;
+
+      // valid dataset URI
+      if (PathUtils.isAncestor(ancestor, new Path(u.getPath()))) {
+        
+        for (URI leaf : validURISet) {
```
Comment (chavdar) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: Current algorithm is a O(n2). You can make it O(nlog n) by a TreeSet for validURISet and exploiting the fact that the necessary condition for a path PF to be a prefix of a path P is to be the largest path smaller that P, i.e.  PF=floor(P).

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+/**
+ * Based on the configuration store to find all {@link ConfigBasedDataset}
+ * @author mitu
+ *
+ */
+
+@Slf4j
+public class ConfigBasedDatasetsFinder implements DatasetsFinder<ConfigBasedDataset> {
+
+  public static final String CONFIG_STORE_ROOT = "config.store.root";
```
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: added

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  public static final String CONFIG_STORE_REPLICATION_ROOT = "config.store.replication.root";
```
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: changed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java (hunk 1 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: changed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java (hunk 3 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: fixed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java (hunk 2 above)
Comment (tuGithub) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java]: fixed

**Pull request merged**